crate_universe/src/metadata.rs

```diff
@@ -319,6 +324,127 @@ impl VendorGenerator {
     }
 }
 
+/// A generate which computes per-platform feature sets.
+pub struct FeatureGenerator {
+    /// The path to a `cargo` binary
+    cargo_bin: PathBuf,
+
+    /// The path to a `rustc` binary
+    rustc_bin: PathBuf,
+}
+
+impl FeatureGenerator {
+    pub fn new(cargo_bin: PathBuf, rustc_bin: PathBuf) -> Self {
+        Self {
+            cargo_bin,
+            rustc_bin,
+        }
+    }
+
+    /// Computes the set of enabled features for each target triplet for each crate.
+    pub fn generate(
+        &self,
+        manifest_path: &Path,
+        platform_triples: &BTreeSet<String>,
+    ) -> Result<BTreeMap<CrateId, SelectList<String>>> {
+        let manifest_dir = manifest_path.parent().unwrap();
+        let mut crate_features = BTreeMap::<CrateId, BTreeMap<String, BTreeSet<String>>>::new();
+        for target in platform_triples {
+            let output = Command::new(&self.cargo_bin)
+                .current_dir(manifest_dir)
+                .arg("tree")
+                .arg("--locked")
+                .arg("--manifest-path")
+                .arg(manifest_path)
+                .arg("--prefix=none")
+                .arg("--format=|{p}|{f}|")
+                .arg("--color=never")
+                .arg("--target")
```

Comment:
Is it possible to pass `all` and parse specifics about each triple from the output?

Reply:
Not obviously. Running with `--target=all` prints the union of all the features enabled by all targets. There doesn't seem to be an option (e.g. for `--format`) to split these out by target.